README.md
My notes/source code while following 'clojure for the brave and true'

http://www.braveclojure.com/

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-01-27
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-01-27
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/brave-clojure-notes/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/brave-clojure-notes/compare/0.1.0...0.1.1


test/brave_clojure_notes/core_test.clj
(ns brave-clojure-notes.core-test
  (:require [clojure.test :refer :all]
            [brave-clojure-notes.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/brave_clojure_notes/chapter_03/hobbit.clj
(ns chapter-notes.hobbit)

(def asym-hobbit-body-parts [{:name "head" :size 3}
                           {:name "left-eye" :size 1}
                           {:name "left-ear" :size 1}
                           {:name "mouth" :size 1}
                           {:name "nose" :size 1}
                           {:name "neck" :size 2}
                           {:name "left-shoulder" :size 3}
                           {:name "left-upper-arm" :size 3}
                           {:name "chest" :size 10}
                           {:name "back" :size 10}
                           {:name "left-forearm" :size 3}
                           {:name "abdomen" :size 6}
                           {:name "left-kidney" :size 1}
                           {:name "left-hand" :size 2}
                           {:name "left-knee" :size 2}
                           {:name "left-thigh" :size 4}
                           {:name "left-lower-leg" :size 3}
                           {:name "left-achilles" :size 1}
                           {:name "left-foot" :size 2}])

(defn matching-part
 [part]
  {:name (clojure
          .string/replace (:name part) #"^left-" "right-")
  :size (:size part)})

; this sucks, but works
(defn symmetrize-body-parts
 "Expects a seq of maps that have a :name and :size"
 [asym-body-parts]
 (loop [remaining-asym-body-parts asym-body-parts
        final-body-parts []]
   (if (empty? remaining-asym-body-parts)
     final-body-parts
   (let [[part & remaining] remaining-asym-body-parts]
     (recur remaining
            (into final-body-parts
                  (set [part (matching-part part)])))))))


(symmetrize-body-parts asym-hobbit-body-parts)

; it's better to use reduce, this is how we could implement it
(defn my-reduce
  ([f initial coll]
   (loop [result initial
          remaining coll]
     (if (empty? remaining)
       result
       (recur (f result (first remaining))(rest remaining)))))
  ([f [head & tail]] ;; variable arity
   (my-reduce f head tail)))

;;; a better way of mirroring body parts using reduce
(defn better-symmetrize-body-parts
  "Expects a seq of maps that hava a :name and :size"
  [asym-body-parts]
  (reduce (fn [final-body-parts part]
            (into final-body-parts (set [part (matching-part part)])))
          []
          asym-body-parts))

(defn hit
  [asym-body-parts]
  (let [sym-parts (better-symmetrize-body-parts asym-body-parts)
        body-part-size-sum (reduce + (map :size sym-parts))
        target (rand body-part-size-sum)
        _ (println (str "target is: "  target))]
    (loop [[part & remaining] sym-parts
           accumulated-size (:size part)]
      (if (> accumulated-size target)
        part
        (recur remaining (+ accumulated-size (:size (first remaining))))))))


(defn mapset
  [f coll]
  (set (map f coll)))


src/brave_clojure_notes/chapter_10/clojure_metaphysics.clj
(ns clojure-metaphysics)

;;---------------------- Atoms ------------------------
;; A reference type for shared, synchronous, independent state.
;; Atoms are free of race conditions.

;; we create an atom (a reference type), give it an identity (fred),
;; and set its initial state
(def fred (atom {:cuddle-hunger-level 0
                 :percent-detoriated  0}))

;; to get an atom's state you dereference it
;; atoms are 'synchronous', meaning we get the value immediately
;; as opposed to getting some time in the future when the value is 'known'
@fred
;; => {:cuddle-hunger-level 0, :percent-detoriated 0}

;; the atom returns an immutable value, which is inherently thread-safe
(let [zombie-state @fred]
  (if (>= (:percent-detoriated zombie-state) 50)
    (future (println (:percent-detoriated zombie-state)))))


(swap! fred
       (fn [current-state]
         (merge-with + current-state {:cuddle-hunger-level 1})))
;; => {:cuddle-hunger-level 1, :percent-detoriated 0}


(defn increase-cuddle-hunger-level
  [zombie-state increase-by]
  (merge-with + zombie-state {:cuddle-hunger-level increase-by}))

;; does not update the atom, because swap! is not called
(increase-cuddle-hunger-level @fred 10)

;; does update the atom, because we are calling swap!
(swap! fred increase-cuddle-hunger-level 10)

;; update-in is a neat little function
;; (update-in <map i want to update> <key path> <update func>)
(update-in {:a {:b 3}} [:a :b] inc)

;; update-in can be used to update the atom
(swap! fred update-in [:cuddle-hunger-level] + 10)

;; we can always retain previous atom states
(let [num (atom 1)
      s1 @num]
  (swap! num inc)
  (println "State 1: " s1)
  (println "Current state:" @num))

;; swap! implements compare-and-set semantics:
;; 1. It reads the current state of the atom.
;; 2. It then applies the update function to that state.
;; 3. It checks whether the previous state has changed meanwhile
;; 4. If not, it updates the atom to refer to the new value
;; 5. If it has, then swap! retries the process, starting over from step 1
;;
;; these semantics are meant to avoid any race conditions
;;
;; updates happen synchronously, so update functions do have the capability of
;; blocking their thread


;; we can also update the atom without checking its current value (overwrite)
(reset! fred {:cuddle-hunger-level 0
              :percent-detoriated  0})


;;---------------------- Watches ------------------------
;; Watches allows you to run a func when the reference reaches a given state

(defn shuffle-speed
  [zombie]
  (* (:cuddle-hunger-level zombie)
     (- 100 (:percent-detoriated zombie))))

;; this is the function to run when a reference is updated
(defn shuffle-alert
  [key watched old-state new-state]
  (let [sph (shuffle-speed new-state)]
    (if (> sph 5000)
      (do
        (println "Run, you fool!")
        (println "The zombie's SPH is now " sph)
        (println "This message brought to your courtesy of " key))
      (do
        (println "All's well with " key)
        (println "Cuddle hunger: " (:cuddle-hunger-level new-state))
        (println "Percent deteriorated: " (:percent-deteriorated new-state))
        (println "SPH" sph)))))


(reset! fred {:cuddle-hunger-level 22
              :percent-detoriated  2})

;; add a watch to fred, the key param can be used to remove the watch later
(add-watch fred :fred-shuffle-alert shuffle-alert)

;; lets update the atom so the watch fn is called
(swap! fred update-in [:percent-detoriated] + 1)
(swap! fred update-in [:cuddle-hunger-level] + 30)

;;---------------------- Validators ------------------------
;; Validators let you specify the legal states of a ref

;; The exception would be thrown anyway, but we add a more descriptive message
(defn percent-detoriated-validator
  [{:keys [percent-detoriated]}]
  (or (and (>= percent-detoriated 0)
           (<= percent-detoriated 100))
      (throw (IllegalStateException. "That's not mathy!"))))

(def bobby
  (atom
    {:cuddle-hunger-level 0 :percent-detoriated 0}
    :validator percent-detoriated-validator))

;; This will throw an IllegalStateException
(swap! bobby update-in [:percent-detoriated] + 200)


;;---------------------- Refs ------------------------
;; Refs are capable of handling synchronization between different identities














src/brave_clojure_notes/chapter_05/functional_programming.clj
(ns brave-clojure.functional-programming)

(defn sum
  ([vals] (sum vals 0))
  ([vals accumulating-total]
   (if (empty? vals)
     accumulating-total
     (sum (rest vals) (+ (first vals) accumulating-total)))))

;;; use recur to acoid StackOverFlow
(defn sum
  ([vals] sum vals 0)
  ([valse accumulating-total]
   (if (empty? vals)
     accumulating-total
     (recur (rest vals) (+ (first vals) accumulating-total)))))


;;; ---------Function composition----------
((comp inc *) 2 7)
;;; => 3

(def character
  {:name "Smooches McCutes"
   :attributes {:intelligence 10
                :strength 4
                :dexterity 5}})
(def c-int (comp :intelligence :attributes))
(def c-str (comp :strength :attributes))
(def c-dex (comp :dexterity :attributes))

(c-int character)
(c-str character)
(c-dex character)

(defn spell-slots
  [char]
  (int (inc (/ (c-int char) 2))))

(spell-slots character)

(def spell-slots-comp (comp int inc #(/ % 2) c-int))

(spell-slots-comp character)

(defn two-comp
  [f g]
  (fn [& args]
    (f (apply g args))))

((two-comp inc inc) 1)

((multi-comp-1 inc inc inc) 1)

(defn multi-comp-2
  [& fns]
  (fn [& args]
    (reduce (fn [result-so-far next-fn] (next-fn result-so-far))
            (apply (last fns) args) (rest (reverse fns)))))

(defn multi-comp-with-let
  [& fns]
  (fn [& args]
    (let [ordered-fns (reverse fns)
          first-result (apply (first ordered-fns) args)
          remaining-fns (rest ordered-fns)]
      (reduce
       (fn [result-so-far next-fn] (next-fn result-so-far))
       first-result
       remaining-fns))))

((multi-comp-with-loop inc inc inc) 1)

(defn multi-comp-with-loop
  [& fns]
  (fn
    [& args]
    (let [ordered-fns (reverse fns)
          first-result (apply (first ordered-fns) args)]
      (loop [result-so-far first-result
             remaining-fns (rest ordered-fns)]
        (if (empty? remaining-fns)
          result-so-far
          (let [next-fn (first remaining-fns)]
            (recur (next-fn result-so-far) (rest remaining-fns))))))))

((multi-comp-with-loop inc inc inc) 1)

;;; -------Memoization------------
(defn sleepy-identity
  [x]
  (Thread/sleep 1000)
  x)

(sleepy-identity "Mr. Fantastico")
(sleepy-identity "Mr. Fantastico")

(def memo-sleep-identity (memoize sleepy-identity))

(memo-sleep-identity "Mr. Fantastico")
; => "Mr. Fantastico" after 1 second

(memo-sleep-identity "Mr. Fantastico")
; => "Mr. Fantastico" immediately



src/brave_clojure_notes/chapter_04/core_functions_in_depth.clj
(ns chapter-notes.core-functions-in-depth)

(def human-consumption   [8.1 7.3 6.6 5.0])
(def critter-consumption [0.0 0.2 0.3 1.1])

(defn unify-diet-data
  [human critter]
  {:human human
   :critter critter})

(map unify-diet-data human-consumption critter-consumption)


;;; passing map a collection of functions
(def sum #(reduce + %))
(def avg #(/ (sum %) (count %)))

(defn stats
  [numbers]
  (map #(% numbers) [sum count avg]))

(stats [3 4 10])

(def identities
  [{:alias "Batman" :real "Bruce Wayne"}
   {:alias "Spider-Man" :real "Peter Parker"}
   {:alias "Santa" :real "Your mom"}
   {:alias "Easter Bunny" :real "Your dad"}])

(map :real identities)

;;; transform a maps' values using reduce
(reduce (fn [new-map [key val]]
          (assoc new-map key (inc val)))
        {}
        {:max 30 :min 10})

;;; filter out keys based on a value
(reduce (fn [new-map [key val]]
          (if (> val 4)
            (assoc new-map key val)
            new-map))
        {}
        {:human 4.1
         :critter 3.9})

(def food-journal
  [{:month 1 :day 1 :human 5.3 :critter 2.3}
   {:month 1 :day 2 :human 5.1 :critter 2.0}
   {:month 2 :day 1 :human 4.9 :critter 2.1}
   {:month 2 :day 2 :human 5.0 :critter 2.5}
   {:month 3 :day 1 :human 4.2 :critter 3.3}
   {:month 3 :day 2 :human 4.0 :critter 3.8}
   {:month 4 :day 1 :human 3.7 :critter 3.9}
   {:month 4 :day 2 :human 3.7 :critter 3.6}])

(take-while #(< (:month %) 3) food-journal)
(drop-while #(< (:month %) 3) food-journal)




(def vampire-database
  {0 {:makes-blood-puns? false, :has-pulse? true  :name "McFishwich"}
   1 {:makes-blood-puns? false, :has-pulse? true  :name "McMackson"}
   2 {:makes-blood-puns? true,  :has-pulse? false :name "Damon Salvatore"}
   3 {:makes-blood-puns? true,  :has-pulse? true  :name "Mickey Mouse"}})

(defn vampire-related-details
  [social-security-number]
  (Thread/sleep 1000)
  (get vampire-database social-security-number))

(defn vampire?
  [record]
  (and (:makes-blood-puns? record)
       (not (:has-pulse? record))
       record))

(defn identify-vampire
  [social-security-numbers]
  (first (filter vampire?
                 (map vampire-related-details social-security-numbers))))



(time (vampire-related-details 0))

(time (def mapped-details (map vampire-related-details (range 0 1000000))))

;; (time (first mapped-details))

;;; infinite sequences
(concat (take 8 (repeat "na")) ["Batman!"])
(take 3 (repeatedly (fn [] (rand-int 10))))

(defn even-numbers
  ([] (even-numbers 0))
  ([n] (cons n (lazy-seq (even-numbers (+ n 2))))))

(take 10 (even-numbers))

;;; will return a seq
(map identity {:sunlight-reaction "Glitter!"})
;;; into will "cast" it back to a map
(into {} (map identity {:sunlight-reaction "Glitter!"}))

;;; conj works differently than into
(conj [0] [1]) ;; => [0 [1]]
(into [0] [1]) ;; => [0 1]

(conj [0] 1 2 3 4)
(conj {:time "midnight"} [:place "ye olde cemetarium"])


;;; (max [1 2 3]) does not work, max expects x & xs
(max 1  2  3)
;;; apply explodes the elements of a collection
(apply max [1 2 3])

(def add10 (partial + 10))
(add10 3) ;; => 13
(add10 5) ;; => 15

(defn my-partial
  [partialized-fn & args]
  (fn [& more-args]
    (apply partialized-fn (into args more-args))))

(def add20 (my-partial + 20))
(add20 3) ;; => 23

(defn lousy-logger
  [log-level message]
  (condp = log-level
    :warn (clojure.string/lower-case message)
    :emergency (clojure.string/upper-case message)))

(def warn (partial lousy-logger :warn))

(warn "Red light ahead")

(defn identify-humans
  [social-security-numbers]
  (filter #(not (vampire? %)) ;; here we could just use complement
          (map vampire-related-details social-security-numbers)))

(def not-vampire? (complement vampire?))

(defn my-complement
  [f]
  (fn [& args]
    (not (apply f args))))

(def my-pos? (complement neg?))
(my-pos? 1)
(my-pos? 2)


src/brave_clojure_notes/10_clojure_metaphysics/atoms_watches_validators.clj
(ns atoms-watches-validators)

;;---------------------- Atoms ------------------------
;; A reference type for shared, synchronous, independent state.
;; Atoms are free of race conditions.

;; we create an atom (a reference type), give it an identity (fred),
;; and set its initial state
(def fred (atom {:cuddle-hunger-level 0
                 :percent-detoriated  0}))

;; to get an atom's state you dereference it
;; atoms are 'synchronous', meaning we get the value immediately
;; as opposed to getting some time in the future when the value is 'known'
@fred
;; => {:cuddle-hunger-level 0, :percent-detoriated 0}

;; the atom returns an immutable value, which is inherently thread-safe
(let [zombie-state @fred]
  (if (>= (:percent-detoriated zombie-state) 50)
    (future (println (:percent-detoriated zombie-state)))))


(swap! fred
       (fn [current-state]
         (merge-with + current-state {:cuddle-hunger-level 1})))
;; => {:cuddle-hunger-level 1, :percent-detoriated 0}


(defn increase-cuddle-hunger-level
  [zombie-state increase-by]
  (merge-with + zombie-state {:cuddle-hunger-level increase-by}))

;; does not update the atom, because swap! is not called
(increase-cuddle-hunger-level @fred 10)

;; does update the atom, because we are calling swap!
(swap! fred increase-cuddle-hunger-level 10)

;; update-in is a neat little function
;; (update-in <map i want to update> <key path> <update func>)
(update-in {:a {:b 3}} [:a :b] inc)

;; update-in can be used to update the atom
(swap! fred update-in [:cuddle-hunger-level] + 10)

;; we can always retain previous atom states
(let [num (atom 1)
      s1 @num]
  (swap! num inc)
  (println "State 1: " s1)
  (println "Current state:" @num))

;; swap! implements compare-and-set semantics:
;; 1. It reads the current state of the atom.
;; 2. It then applies the update function to that state.
;; 3. It checks whether the previous state has changed meanwhile
;; 4. If not, it updates the atom to refer to the new value
;; 5. If it has, then swap! retries the process, starting over from step 1
;;
;; these semantics are meant to avoid any race conditions
;;
;; updates happen synchronously, so update functions do have the capability of
;; blocking their thread


;; we can also update the atom without checking its current value (overwrite)
(reset! fred {:cuddle-hunger-level 0
              :percent-detoriated  0})


;;---------------------- Watches ------------------------
;; Watches allows you to run a func when the reference reaches a given state

(defn shuffle-speed
  [zombie]
  (* (:cuddle-hunger-level zombie)
     (- 100 (:percent-detoriated zombie))))

;; this is the function to run when a reference is updated
(defn shuffle-alert
  [key watched old-state new-state]
  (let [sph (shuffle-speed new-state)]
    (if (> sph 5000)
      (do
        (println "Run, you fool!")
        (println "The zombie's SPH is now " sph)
        (println "This message brought to your courtesy of " key))
      (do
        (println "All's well with " key)
        (println "Cuddle hunger: " (:cuddle-hunger-level new-state))
        (println "Percent deteriorated: " (:percent-deteriorated new-state))
        (println "SPH" sph)))))


(reset! fred {:cuddle-hunger-level 22
              :percent-detoriated  2})

;; add a watch to fred, the key param can be used to remove the watch later
(add-watch fred :fred-shuffle-alert shuffle-alert)

;; lets update the atom so the watch fn is called
(swap! fred update-in [:percent-detoriated] + 1)
(swap! fred update-in [:cuddle-hunger-level] + 30)

;;---------------------- Validators ------------------------
;; Validators let you specify the legal states of a ref

;; The exception would be thrown anyway, but we add a more descriptive message
(defn percent-detoriated-validator
  [{:keys [percent-detoriated]}]
  (or (and (>= percent-detoriated 0)
           (<= percent-detoriated 100))
      (throw (IllegalStateException. "That's not mathy!"))))

(def bobby
  (atom
    {:cuddle-hunger-level 0 :percent-detoriated 0}
    :validator percent-detoriated-validator))

;; This will throw an IllegalStateException
(swap! bobby update-in [:percent-detoriated] + 200)


;;---------------------- Refs ------------------------
;; Refs are capable of handling synchronization between different 'identities'












